x/oracle/keeper/historic_price.go

```diff
@@ -0,0 +1,224 @@
+			Cmp(prices[j].ExchangeRate.BigInt()) > 0
+	})
+
+	if lenPrices%2 == 0 {
+		return prices[lenPrices/2-1].ExchangeRate.
+			Add(prices[lenPrices/2].ExchangeRate).
+			QuoInt64(2)
+	}
+	return prices[lenPrices/2].ExchangeRate
+}
+
+// medianDeviation returns the standard deviation around the
+// median of a list of prices.
+// medianDeviation = ∑((price - median)^2 / len(prices))
+func medianDeviation(median sdk.Dec, prices []types.HistoricPrice) sdk.Dec {
+	lenPrices := len(prices)
+	medianDeviation := sdk.ZeroDec()
+
+	for _, price := range prices {
+		medianDeviation = medianDeviation.Add(price.ExchangeRate.
+			Sub(median).Abs().Power(2).
+			QuoInt64(int64(lenPrices)))
+	}
+
+	return medianDeviation
+}
+
+// GetMedian returns a given denom's median price in the last prune
+// period since a given block.
+func (k Keeper) GetMedian(
+	ctx sdk.Context,
+	denom string,
+	blockNum uint64,
+) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetMedianKey(denom, blockNum))
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrUnknownDenom, denom)
+	}
+
+	decProto := sdk.DecProto{}
+	k.cdc.MustUnmarshal(bz, &decProto)
+
+	return decProto.Dec, nil
+}
+
+// SetMedian uses all the historic prices of a given denom to calculate
+// its median price in the last prune period since the current block and
+// set it to the store. It will also call setMedianDeviation with the
+// calculated median.
+func (k Keeper) SetMedian(
+	ctx sdk.Context,
+	denom string,
+) {
+	store := ctx.KVStore(k.storeKey)
+	historicPrices := k.GetHistoricPrices(ctx, denom)
+	median := median(historicPrices)
+	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: median})
+	store.Set(types.GetMedianKey(denom, uint64(ctx.BlockHeight())), bz)
+	k.setMedianDeviation(ctx, denom, median, historicPrices)
+}
+
+// GetMedianDeviation returns a given denom's standard deviation around
+// its median price in the last prune period since a given block.
+func (k Keeper) GetMedianDeviation(
+	ctx sdk.Context,
+	denom string,
+	blockNum uint64,
+) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetMedianDeviationKey(denom, blockNum))
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrUnknownDenom, denom)
+	}
+
+	decProto := sdk.DecProto{}
+	k.cdc.MustUnmarshal(bz, &decProto)
+
+	return decProto.Dec, nil
+}
+
+// setMedianDeviation sets a given denom's standard deviation around
+// its median price in the last prune period since the current block.
+func (k Keeper) setMedianDeviation(
+	ctx sdk.Context,
+	denom string,
+	median sdk.Dec,
+	prices []types.HistoricPrice,
+) {
+	store := ctx.KVStore(k.storeKey)
+	medianDeviation := medianDeviation(median, prices)
+	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: medianDeviation})
+	store.Set(types.GetMedianDeviationKey(denom, uint64(ctx.BlockHeight())), bz)
+}
+
+// GetHistoricPrice returns the historic price of a denom at a given
+// block.
+func (k Keeper) GetHistoricPrice(
```

Comment:
Does this need to be a public keeper function?
```suggestion
func (k Keeper) getHistoricPrice(
```

Reply:
This method isn't called within the keeper so since it doesn't need to be public I've removed it